apps/emqx_resource/src/emqx_resource.erl

```diff
@@ -486,6 +492,30 @@ apply_reply_fun({F, A}, Result) when is_function(F) ->
 apply_reply_fun(From, Result) ->
     gen_server:reply(From, Result).
 
+-spec allocate_resource(resource_id(), any(), term()) -> ok.
+allocate_resource(InstanceId, Key, Value) ->
+    true = ets:insert(?RESOURCE_ALLOCATION_TAB, {InstanceId, Key, Value}),
+    ok.
+
+-spec has_allocated_resources(resource_id()) -> boolean().
+has_allocated_resources(InstanceId) ->
+    case ets:lookup(?RESOURCE_ALLOCATION_TAB, InstanceId) of
+        [] ->
+            false;
+        [_ | _] ->
+            true
+    end.
+
+-spec get_allocated_resources(resource_id()) -> map().
+get_allocated_resources(InstanceId) ->
+    Objects = ets:lookup(?RESOURCE_ALLOCATION_TAB, InstanceId),
+    maps:from_list([{K, V} || {_InstanceId, K, V} <- Objects]).
+
+-spec delete_allocated_resources(resource_id()) -> ok.
+delete_allocated_resources(InstanceId) ->
```

Comment:
should the name of the function take into account that resources are not being actually released here?

Reply:
Good point.  I'll rename it to `forget_allocated_resources`, as the actual freeing of resources cannot be done here.